chore(ZNTA-107) add build script

seanf opened the pull request:
JIRA issue URL: https://zanata.atlassian.net/browse/ZNTA-107

Create build script for Zanata


## Reviewer tasks

The following is based on the
[Zanata Development Guidelines](https://github.com/zanata/zanata-server/wiki/Development-Guidelines).
As a reviewer, you should check the guidelines regularly for updates, and just
to refresh your memory.


The reviewer can use this list for reference to make sure
they have considered all the important points.

Use the checkboxes or not, whatever works best for you.

 - Code quality
   - [ ] Code passes style check (checkstyle/eslint)
   - [ ] Code is clear and concise
   - [ ] UI code is internationalised
   - [ ] Code has adequate comments where appropriate
   - Code is in an appropriate place
     - [ ] Classes are in appropriate packages
     - [ ] Code fits with class's responsibilities (SRP)
   - [ ] Configuration files are documented enough
   - If code is removed
     - [ ] No remaining code references the removed code
       - [ ] No orphaned rewrite rules
       - [ ] No orphaned config settings
     - [ ] No remaining comments refer to the removed code
     - [ ] Unused imports and libraries are removed
 - Testing
   - [ ] New code has tests
   - [ ] Changed code has tests
   - [ ] All tests pass
   - [ ] Test coverage stable or increased
 - Documentation
   - [ ] New features are documented
   - [ ] Docs removed for deleted features
   - [ ] Docs updated for all changed features
   - [ ] Developer/sysadmin docs updated if appropriate
 - If there are new dependencies
   - [ ] Does each new dependency pass all the
         [Considerations for new dependencies](https://github.com/zanata/zanata-server/wiki/Development-Guidelines#new-technologies-and-dependencies)?


----
*This template can be updated in .github/PULL_REQUEST_TEMPLATE.md*

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/zanata/zanata-platform/75)
<!-- Reviewable:end -->

aeng commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> I can't spot anything for now but I think we probably will when we start using it.

---

Review status: 0 of 1 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/75#-:-KXbpvZPo53oFDs_Z2GZ:bz7t96w)*
<!-- Sent from Reviewable.io -->

djansen-redhat commented:
@seanf Should we delete quickbuild as part of this, or nah?

aeng commented:
@djansen-redhat I think we should to avoid maintenance of multiple scripts that's doing the same thing.

seanf commented:
@aeng Could you review my changes please? Once I removed `quickbuild.sh`, I had to update the readme, then I had to fix the rest of the readme, then I ended up replacing `cargowait.sh` instead of explaining why it won't work. I also fixed a couple of warts like packaging everything twice when using `-i`.

huangp commented:
I tried it and here is my experience:
- [ ] now the build script is at platform level so it will build *every thing*. I know we have discussed this but still sometimes I will prefer just build the server. p.s. it failed on building client for me when I run build -w.
- [ ] dry run output didn't escape !. So you can't just copy and paste to run it
- [ ] -w still runs frontend and gwt-editor by default even it says quick build

seanf commented:
1. Have you tried using ./build.sh --server (from root), or ../build.sh .
(from server/)?

2. I think the best we can do there is output a warning that escaping may
be required. (DONE.) But contributions welcome!

3. It's as quick as it can be without compromising
correctness/up-to-date-ness of the functional tests. "Quick" in the build
script help means no unit tests.

PS please try to use long options when writing scripts which call this
script, or talking to other people about it. "--wait" communicates more
than "-w".


On 13 December 2016 at 10:06, Patrick Huang <notifications@github.com>
wrote:

> I tried it and here is my experience:
>
>    - now the build script is at platform level so it will build *every
>    thing*. I know we have discussed this but still sometimes I will
>    prefer just build the server. p.s. it failed on building client for me when
>    I run build -w.
>    - dry run output didn't escape !. So you can't just copy and paste to
>    run it
>    - -w still runs frontend and gwt-editor by default even it says quick
>    build
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zanata/zanata-platform/pull/75#issuecomment-266594063>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AADTiyWbtKSIuhui_78rKjjfWWjxQkBNks5rHeGQgaJpZM4K7STn>
> .
>



-- 
Sean Flanigan

Principal Software Engineer
Globalisation Tools Engineering
Red Hat

huangp commented:
1. now the first option but running from server/ will result in mvnw not found error
2. 3. 4. ok

seanf commented:
@aeng I can haz review?![](http://cdn.meme.am/instances/500x/55070035.jpg)

carlosmunoz commented:
@seanf I'm glad you're jumping on the meme bandwagon. What made you change your mind? 😉

seanf commented:
@carlosmunoz there's a blanket exemption for cat pictures. Didn't you know?

seanf commented:
@zanata-jenkins test this please.

aeng commented:
I know we not using profile anymore and mostly targetting docker run, but for those who is still using local server, is it possible to have -Pexplode too?

---

Reviewed 1 of 1 files at r1, 2 of 6 files at r2, 1 of 1 files at r5.
Review status: 3 of 6 files reviewed at latest revision, 1 unresolved discussion.

---

*[README.md, line 42 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/75#-KZ3WpbKxhnudx23a4YE:-KZ3WpbKxhnudx23a4YF:bp5cgx1) ([raw file](https://github.com/zanata/zanata-platform/blob/7e6771da192cc209ea6f774ba6ff8eb4cfd3b969/README.md#L42)):*
> ```Markdown
> - JBoss EAP 7 or WildFly 10 (optional)
> - Linux or Mac OSX. Windows works too, sometimes.
> ```

maven too?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/75#-:-KZ3Z5gDTf-Uxzf4YwRz:b325spi)*
<!-- Sent from Reviewable.io -->

seanf commented:
Any unrecognised options will be passed through to mvn. Just put it at the end. If I were you, I would try to use -Dsomething, not -Psomething. The -P option in Maven doesn't really play well with other profiles.

---

Review status: 3 of 6 files reviewed at latest revision, 1 unresolved discussion.

---

*[README.md, line 42 at r5](https://reviewable.io:443/reviews/zanata/zanata-platform/75#-KZ3WpbKxhnudx23a4YE:-KZ4-rBwNTJNWowSZoVB:b-ve31pt) ([raw file](https://github.com/zanata/zanata-platform/blob/7e6771da192cc209ea6f774ba6ff8eb4cfd3b969/README.md#L42)):*
<details><summary><i>Previously, aeng (Alex Eng) wrote…</i></summary><blockquote>

maven too?
</blockquote></details>

We use the Maven wrapper now, so no, you don't have to install Maven yourself. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/75#-:-KZ407nHmUiVOWruaT1p:b-xt2m9s)*
<!-- Sent from Reviewable.io -->

aeng commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 6 files at r2.
Review status: 4 of 6 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/75#-:-KZ410bWXJnohnnCj-DH:bnfp4nl)*
<!-- Sent from Reviewable.io -->

spathare21 commented:
Verified ./build from root dir with all possible options and it works as expected
